File: 26S/session8.py
# session8.py
# Session 8: Instrumental Variables — 2SLS, weak-IV diagnostics, LATE
# End-to-end on Card (1995) "Using Geographic Variation in College Proximity"
# — the canonical IV teaching dataset.
#
#   Outcome Y    : lwage  (log hourly wage)
#   Treatment D  : educ   (years of schooling, endogenous)
#   Instrument Z : nearc4 (lives near a 4-year college)
#
# Required packages:
#   pip install causaldata pandas numpy statsmodels linearmodels

import numpy as np
import pandas as pd
import statsmodels.api as sm
from linearmodels.iv import IV2SLS
from causaldata import close_college      # Card (1995) college proximity

df = close_college.load_pandas().data.dropna().reset_index(drop=True).copy()
# Cast `exper` to float64 BEFORE squaring. `causaldata.close_college` ships
# `exper` as int8; for exper >= 12 the squared value silently overflows the
# int8 range [-128, 127] (e.g. 25**2 = 625 wraps to 113), which biases every
# downstream regression. R does not have this problem because it represents
# integers as 32-bit by default.
df["exper"]  = df["exper"].astype(float)
df["exper2"] = df["exper"] ** 2
# `causaldata.close_college` is the trimmed teaching slice of Card (1995):
# it drops the regional dummies (reg661-reg668), so we use the basic controls.
controls = ["exper", "exper2", "black", "south", "smsa", "married"]

# ============================================================
# 1. Naive OLS (likely biased: ability confounds educ and wage)
# ============================================================
X_ols = sm.add_constant(df[["educ"] + controls])
m_ols = sm.OLS(df["lwage"], X_ols).fit(cov_type="HC1")
print("---- OLS (likely biased) ----")
print(f"  beta_educ = {m_ols.params['educ']:+.4f}  "
      f"(SE = {m_ols.bse['educ']:.4f})")

# ============================================================
# 2. Reduced form: does the instrument move the outcome?
# ============================================================
X_rf = sm.add_constant(df[["nearc4"] + controls])
m_rf = sm.OLS(df["lwage"], X_rf).fit(cov_type="HC1")
print("\n---- Reduced form (Y on Z) ----")
print(f"  beta_Z    = {m_rf.params['nearc4']:+.4f}  "
      f"(SE = {m_rf.bse['nearc4']:.4f})")

# ============================================================
# 3. First stage: does the instrument move the treatment?
#    Rule of thumb: F > 10 (Stock-Yogo) for a single IV.
# ============================================================
X_fs = sm.add_constant(df[["nearc4"] + controls])
m_fs = sm.OLS(df["educ"], X_fs).fit(cov_type="HC1")
print("\n---- First stage (D on Z) ----")
print(f"  pi_Z      = {m_fs.params['nearc4']:+.4f}  "
      f"(SE = {m_fs.bse['nearc4']:.4f})")
# Robust F for the single excluded instrument:
F = (m_fs.params["nearc4"] / m_fs.bse["nearc4"]) ** 2
print(f"  First-stage F (single Z) = {F:.2f}   (need > 10)")

# ============================================================
# 4. 2SLS via linearmodels.iv.IV2SLS
# ============================================================
exog  = sm.add_constant(df[controls])
endog = df[["educ"]]
instr = df[["nearc4"]]
y     = df["lwage"]

m_iv = IV2SLS(y, exog, endog, instr).fit(cov_type="robust")
print("\n---- 2SLS estimate (LATE on compliers) ----")
print(m_iv.summary)
# Useful diagnostics:
print("\nFirst-stage diagnostics:")
print(m_iv.first_stage)
# m_iv.wu_hausman()      # endogeneity test (OLS vs IV)
# m_iv.sargan            # only meaningful when overidentified

# ============================================================
# 5. Wald estimator: ratio of reduced-form to first-stage coefficients
#    (matches 2SLS exactly with one instrument and no covariates)
# ============================================================
beta_rf = sm.OLS(df["lwage"], sm.add_constant(df["nearc4"])).fit().params["nearc4"]
pi_fs   = sm.OLS(df["educ"],  sm.add_constant(df["nearc4"])).fit().params["nearc4"]
print(f"\n---- Wald estimator (no covariates): {beta_rf / pi_fs:+.3f} ----")

# ============================================================
# 6. What to report:
#   * 2SLS coefficient on `educ` -> the LATE (effect for compliers,
#     i.e. those induced into more schooling by living near a college)
#   * Robust SE / 95% CI
#   * First-stage F (relevance)
#   * Wu-Hausman p (endogeneity)
#   * If overidentified: Sargan p (partial exclusion check)
# ============================================================

File: 26S/session8.R
# session8.R
# Session 8: Instrumental Variables — 2SLS, weak-IV diagnostics, LATE
# End-to-end on Card (1995) "Using Geographic Variation in College Proximity"
# — the canonical IV teaching dataset.
#
#   Outcome Y    : lwage  (log hourly wage)
#   Treatment D  : educ   (years of schooling, endogenous)
#   Instrument Z : nearc4 (lives near a 4-year college)
#
# Required packages:
#   install.packages(c("causaldata", "AER"))   # AER provides ivreg() directly

library(causaldata)
library(AER)

# `causaldata::close_college` is the canonical Card (1995) college-proximity
# teaching dataset. The trimmed teaching slice drops the regional dummies
# (reg661-reg668), so we use the basic controls only.
card <- na.omit(close_college)
controls <- "exper + I(exper^2) + black + south + smsa + married"

# ============================================================
# 1. Naive OLS (likely biased: ability confounds educ and wage)
# ============================================================
f_ols <- as.formula(paste("lwage ~ educ +", controls))
m_ols <- lm(f_ols, data = card)
cat("---- OLS (likely biased) ----\n")
print(coef(summary(m_ols))["educ", ])

# ============================================================
# 2. Reduced form: does the instrument move the outcome?
# ============================================================
f_rf  <- as.formula(paste("lwage ~ nearc4 +", controls))
m_rf  <- lm(f_rf, data = card)
cat("\n---- Reduced form (Y on Z) ----\n")
print(coef(summary(m_rf))["nearc4", ])

# ============================================================
# 3. First stage: does the instrument move the treatment?
#    Rule of thumb: F > 10 (Stock-Yogo) for a single IV.
# ============================================================
f_fs  <- as.formula(paste("educ ~ nearc4 +", controls))
m_fs  <- lm(f_fs, data = card)
cat("\n---- First stage (D on Z) ----\n")
print(coef(summary(m_fs))["nearc4", ])
cat(sprintf("  First-stage F-stat: %.2f  (need > 10)\n",
            summary(m_fs)$fstatistic[1]))

# ============================================================
# 4. 2SLS via AER::ivreg
# ============================================================
f_iv <- as.formula(paste("lwage ~ educ +", controls,
                         "|", "nearc4 +", controls))
m_iv <- ivreg(f_iv, data = card)
cat("\n---- 2SLS estimate (LATE on compliers) ----\n")
print(summary(m_iv, diagnostics = TRUE))
# Diagnostics block reports:
#  * Weak instruments  (= first-stage F): want LARGE
#  * Wu-Hausman        (OLS vs IV): tests endogeneity
#  * Sargan            (only with overidentification): partial exclusion check

# ============================================================
# 5. Wald estimator: ratio of reduced-form to first-stage coefficients
#    (matches 2SLS exactly with one instrument and no covariates;
#     useful for didactic intuition.)
# ============================================================
m_rf0 <- lm(lwage ~ nearc4, data = card)
m_fs0 <- lm(educ  ~ nearc4, data = card)
wald  <- coef(m_rf0)["nearc4"] / coef(m_fs0)["nearc4"]
cat(sprintf("\n---- Wald estimator (no covariates): %.3f ----\n", wald))

# ============================================================
# 6. Reading the output (what to report)
#   * 2SLS coefficient on `educ` -> the LATE (effect for compliers,
#     i.e. those induced into more schooling by living near a college)
#   * 95% CI / robust SE
#   * First-stage F-stat (relevance check)
#   * Wu-Hausman p-value (endogeneity)
#   * If overidentified: Sargan p-value (partial exclusion check)
# ============================================================
